src/cloud-providers/gcp/provider.go

```diff
@@ -0,0 +1,191 @@
+		}
+	} else {
+		var err error
+		provider.instancesClient, err = compute.NewInstancesRESTClient(context.TODO())
+		if err != nil {
+			return nil, fmt.Errorf("NewInstancesRESTClient error: %s", err)
+		}
+	}
+	return provider, nil
+}
+
+func getIPs(instance *computepb.Instance) ([]netip.Addr, error) {
+	var podNodeIPs []netip.Addr
+	for _, nic := range instance.GetNetworkInterfaces() {
+		for _, access := range nic.GetAccessConfigs() {
+			ipStr := access.GetNatIP()
+			ip, err := netip.ParseAddr(ipStr)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse pod node IP %q: %w", ipStr, err)
+			}
+			podNodeIPs = append(podNodeIPs, ip)
+			logger.Printf("%s", ip.String())
+		}
+	}
+	return podNodeIPs, nil
+}
+
+func (p *gcpProvider) CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator, spec provider.InstanceTypeSpec) (*provider.Instance, error) {
+
+	instanceName := util.GenerateInstanceName(podName, sandboxID, maxInstanceNameLen)
+	logger.Printf("CreateInstance: name: %q", instanceName)
+
+	userData, err := cloudConfig.Generate()
+	if err != nil {
+		return nil, err
+	}
+
+	//Convert userData to base64
+	userDataEnc := base64.StdEncoding.EncodeToString([]byte(userData))
+	logger.Printf("userDataEnc:  %s", userDataEnc)
+
+	insertReq := &computepb.InsertInstanceRequest{
+		Project: p.serviceConfig.ProjectId,
+		Zone:    p.serviceConfig.Zone,
+		InstanceResource: &computepb.Instance{
+			Name: proto.String(instanceName),
+			Disks: []*computepb.AttachedDisk{
+				{
+					InitializeParams: &computepb.AttachedDiskInitializeParams{
+						DiskSizeGb:  proto.Int64(20),
+						SourceImage: proto.String(fmt.Sprintf("projects/%s/global/images/%s", p.serviceConfig.ProjectId, p.serviceConfig.ImageName)),
+						DiskType:    proto.String(fmt.Sprintf("zones/%s/diskTypes/pd-standard", p.serviceConfig.Zone)),
+					},
+					AutoDelete: proto.Bool(true),
+					Boot:       proto.Bool(true),
+					Type:       proto.String(computepb.AttachedDisk_PERSISTENT.String()),
+				},
+			},
+			Metadata: &computepb.Metadata{
+				Items: []*computepb.Items{
+					{
+						Key:   proto.String("user-data"),
+						Value: proto.String(userDataEnc),
+					},
+					{
+						Key:   proto.String("user-data-encoding"),
+						Value: proto.String("base64"),
+					},
+				},
+			},
+			MachineType: proto.String(fmt.Sprintf("zones/%s/machineTypes/%s", p.serviceConfig.Zone, p.serviceConfig.MachineType)),
+			NetworkInterfaces: []*computepb.NetworkInterface{
+				{
+					AccessConfigs: []*computepb.AccessConfig{
+						{
+							Name:        proto.String("External NAT"),
+							NetworkTier: proto.String("PREMIUM"),
+						},
+					},
+					StackType: proto.String("IPV4_Only"),
+					Name:      proto.String(p.serviceConfig.Network),
+				},
+			},
+		},
+	}
+	op, err := p.instancesClient.Insert(ctx, insertReq)
+	if err != nil {
+		return nil, fmt.Errorf("Instances.Insert error: %s. req: %v", err, insertReq)
```

Comment:
In case `Insert()` fails, will it automatically delete any resources (e.g. disks and networks) that were created? Asking because Azure does need to delete those resources explicitly, whereas other providers don't.

Reply:
I tried to find any reference or documentation and could not find yet. I will keep looking, however:

Network is not created here, only attached, we are using the default one, for now. (To be improved).

With disks I tried simulating several possible situations that would fail, and the boot disk was not there. So I would say it is safe for now.

Reply:
OK. Can be improved and/or fixed later.